fix(prefect-server): better support for internal and external database configs

mitchnielsen opened the pull request:
## Summary

**This will be a breaking change**, so we should release this in a new major version. The configuration keys have moved around. If we _really_ need to, we can try to build in some backwards-compatibility, but that always ends up looking pretty rough in Helm.

Ensures that the PostgreSQL Secret is created, even if `postgresql.enabled=false`. This ensures that we support a use case where folks want to use an external instance of PostgreSQL, but still want the Secret to automatically be generated with the proper connection string.

With a recent change, we would skip creation of this secret if PostgreSQL was disabled which forced users to create a Secret themselves. This change now allows them to continue providing the `auth` values and letting the chart build the Secret with the correct connection string.

I ended up needing to go a step further by removing any custom configuration that was injected under the `postgresql` key, which should be dedicated to overriding default values from the PostgreSQL chart. There's now a new `secret` key, which aligns with the other top-level keys (`deployment`, `service`, etc.). These values will be considered if `postgresql.enabled=false`.

Closes https://github.com/PrefectHQ/prefect-helm/issues/358

Follow-up to https://github.com/PrefectHQ/prefect-helm/pull/341.

## Testing

See the [added test suite](https://github.com/PrefectHQ/prefect-helm/blob/db-secret-config/charts/prefect-server/tests/database_test.yaml).

mitchnielsen committed:
Create PostgreSQL Secret even if postgresql.enabled=false

Ensures that the PostgreSQL Secret is created, even if
`postgresql.enabled=false`. This ensures that we support a use case
where folks want to use an external instance of PostgreSQL, but still
want the Secret to automatically be generated with the proper connection
string.

With a recent change, we would skip creation of this secret if
PostgreSQL was disabled which forced users to create a Secret
themselves. This change now allows them to continue providing the `auth`
values and letting the chart build the Secret with the correct
connection string.

Closes https://github.com/PrefectHQ/prefect-helm/issues/358

mitchnielsen commented:
@ahgraber @tmyhu @MRocholl - I _believe_ this should cover all the use cases you mentioned. Could you please give this a look when you have a moment and let me know if this needs any tweaking?

Trying to make it flexible enough to cover all use cases while keeping it as consistent/straightforward as possible. Thanks!

tmyhu commented:
@mitchnielsen it looks good to me, but won't make any difference for our use case since we are using SQLite. I assume that's not something you really want to support specifically since there is no hint about it anywhere in the helm chart I could see? We should probably migrate to PostgreSQL anyway, but for now I'm happy to continue creating the secret and connection string manually for SQLite.

ahgraber commented:
@mitchnielsen Thanks for this effort! 

The output connection string from the test deployment's secret is missing the `hostname` in the database address; will that still be pulled in from `postgresql.externalHostname`?  If so, I believe this covers the requested use case.

mitchnielsen commented:
> it looks good to me, but won't make any difference for our use case since we are using SQLite. I assume that's not something you really want to support specifically since there is no hint about it anywhere in the helm chart I could see? We should probably migrate to PostgreSQL anyway, but for now I'm happy to continue creating the secret and connection string manually for SQLite.

@tmyhu fair point. We have this in mind and are tracking in https://github.com/PrefectHQ/prefect-helm/issues/345. I'd like to provide the `auth` configuration outside of the context of the `postgresql` key to make the configuration more clear, and open the door for supporting SQLite since we provide that information in our self-hosted docs. That said, moving to PostgreSQL is probably the ideal longer-term move.

mitchnielsen commented:
@ahgraber great call-out, let me see if I can make a minimal change to make that configuration supported. If not, may need to jump ahead to providing this config outside of the `postgresql` key because of how some of the nested templates are used.

mitchnielsen committed:
Remove custom config from postgresql subchart

The `postgresql` key in `values.yaml` is for overriding the PostgreSQL
chart's default values. It's not a good place to define custom keys
because these are conflated with the actual configuration from the
subchart. This changes removes those custom configs.

It also removes any configured values that already matched the default
from the PostgreSQL subchart, and were therefore not doing anything.

mitchnielsen committed:
Support providing external connection string

When the PostgreSQL subchart is disabled, we need a way to provide the
connection string information for the external instance. This adds the
required values and template helpers to calculate the connection string
in this scenario.

mitchnielsen committed:
Support setting name and creation of Secret

- Supports setting the name to use for the Secret
- Supports controlling whether or not to create the Secret
- Supports someone providing their own existing Secret

mitchnielsen committed:
Add helm unit tests for database configuration

Uses https://github.com/helm-unittest/helm-unittest
to unit test the database configuration. This is a replacement
for coming up with test cases locally, running `helm template`, and
manually validating the output.

mitchnielsen committed:
Add script and CI workflow for helm-unittest

- Adds script to run helm-unittest locally
- Adds CI workflow to run helm-unittest in GitHub Actions

mitchnielsen committed:
Update README for new external db configuration

Updates the README to reflect the new configuration for external
databases.

mitchnielsen commented:
@ahgraber @tmyhu @MRocholl - alrighty, bit of a rework here to make sure we have proper support for either internal or external database configuration. For some examples of the configuration you'd provide, check out the updated docs or the [unit test file](https://github.com/PrefectHQ/prefect-helm/blob/0f43bfc8f082caad1b2faf19a0ef74943ff345e4/charts/prefect-server/tests/database_test.yaml). Welcoming any feedback here 🤝

MRocholl commented:
@mitchnielsen sorry for the late reply. These changes should satisfy the requirements I believe 👍  Thx for the great work

mitchnielsen committed:
Merge branch 'main' into db-secret-config

mitchnielsen commented:
> sorry for the late reply. These changes should satisfy the requirements I believe 👍 Thx for the great work

@MRocholl - no worries at all, thank you for confirming 🤝

parkedwards commented on charts/prefect-server/values.yaml:
```diff
@@ -340,8 +342,6 @@ postgresql:
       # -- enable PostgreSQL Primary data persistence using PVC
       enabled: false
-      # -- PVC Storage Request for PostgreSQL volume
-      size: 8Gi
```
does this default need to be re-implemented somewhere?

mitchnielsen replied:
Shouldn't need to, since it's [the default in Bitnami's values.yaml](https://github.com/bitnami/charts/blob/postgresql/12.12.10/bitnami/postgresql/values.yaml#L685). We could technically leave this here, but I deleted it to make it more clear that everything defined under `postgresql` is an override for the chart's defaults.

parkedwards commented on charts/prefect-server/values.yaml:
```diff
@@ -293,7 +293,26 @@ ingress:
 
-# Postgresql configuration
+# Secret configuration
+secret:
```
just so i understand, `.Values.secret` is specifically for BYO postgres credentials?

mitchnielsen replied:
Exactly, it gives us somewhere outside of the context of `postgresql.` to define database configs (so we can dedicate `postgresql.` settings for the PostgreSQL chart without overloading it with our own chart's configs).

mitchnielsen committed:
Add helm/chart-testing to mise

Installs the helm/chart-testing binary at its latest version.
Needed for the tests defined under `charts/<chart>/tests/test-*.yaml`.

mitchnielsen committed:
Mention how to run the `helm-unittest` helper

mitchnielsen committed:
Merge branch 'main' into db-secret-config
